sbin/mount/getmntopts.3
.\"	BSDI getmntopts.3,v 2.1 1995/02/03 07:21:59 polk Exp
.\"
.\" Copyright (c) 1994
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)getmntopts.3	8.1 (Berkeley) 3/27/94
.\"
.Dd January 10, 1995
.Dt GETMNTOPTS 3
.Os BSD 4.4
.Sh NAME
.Nm getmntopts
.Nd scan mount options
.Sh SYNOPSIS
.Fd #include <mntopts.h>
.Ft void
.Fn getmntopts "char *options" "struct mntopt *mopts" "int *flagp"
.Sh DESCRIPTION
The
.Nm getmntopts
function takes a comma separated option list and a list
of valid option names, and computes the bitmask
corresponding to the requested set of options.
.Pp
The string
.Dv options
is broken down into a sequence of comma separated tokens.
Each token is looked up in the table described by
.Dv mopts
and the bits in
the word referenced by
.Dv flagp
are updated.
The flag word is not initialized by
.Nm getmntopt .
The table,
.Dv mopts ,
has the following format:
.Bd -literal
struct mntopt {
	char *m_option;		/* option name */
	int m_inverse;		/* is this a negative option, eg "dev" */
	int m_flag;		/* bit to set, eg MNT_RDONLY */
	char **m_value;		/* where to put the value */
	int *m_flagp;		/* where to put the flag */
};
.Ed
.Pp
The members of this structure are defined as follows:
.Bl -tag -width m_inverse
.It Fa m_option
is the option name,
for example
.Dq suid .
.It Fa m_inverse
tells
.Nm getmntopts
that the name has the inverse meaning of the
bit.
For example,
.Dq suid
is the string, whereas the
mount flag is
.Dv MNT_NOSUID .
In this case, the sense of the string and the flag
are inverted, so the
.Dv m_inverse
flag should be set.
.It Fa m_flag
is the value of the bit to be set or cleared in
the flag word when the option is recognized.
The bit is set when the option is discovered,
but cleared if the option name was preceded
by the letters
.Dq no .
The
.Dv m_inverse
flag causes these two operations to be reversed.
.It Fa m_value
indicates that a value may be specified by appending
.Dv =value
to the option.
The
.Dq no
prefix does not apply when a value is specified.
A pointer to the value is placed in
.Fa *m_value .
.It Fa m_flagp
when not
.Dq NULL ,
points to an alternate flag word to be used instead of the
flag word pointed to by
.Fa flagp .
.El
.Pp
Each of the user visible
.Dv MNT_
flags has a corresponding
.Dv MOPT_
macro which defines an appropriate
.Li "struct mntopt"
entry.
To simplify the program interface and ensure consistency across all
programs, a general purpose macro,
.Dv MOPT_STDOPTS ,
is defined which
contains an entry for all the generic VFS options.
In addition, the macros
.Dv MOPT_FORCE
and
.Dv MOPT_UPDATE
exist to enable the
.Dv MNT_FORCE
and
.Dv MNT_UPDATE
flags to be set.
Finally, the table must be terminated by an entry with a NULL
first element.
.Sh EXAMPLES
Most commands will use the standard option set.
Local filesystems which support the
.Dv MNT_UPDATE
flag, would also have an
.Dv MOPT_UPDATE
entry.
This can be declared and used as follows:
.Bd -literal
#include "mntopts.h"

struct mntopt mopts[] = {
	MOPT_STDOPTS,
	MOPT_UPDATE,
	{ NULL }
};

	...
	mntflags = 0;
	...
	getmntopts(options, mopts, &mntflags)
	...
.Ed
.Sh DIAGNOSTICS
The
.Nm getmntopts
function displays an error message and exits if an
unrecognized option is encountered.
.Sh SEE ALSO
.Xr err 3 ,
.Xr mount 8
.Sh HISTORY
The
.Fn getmntopts
function appeared in 
.Bx 4.4 .


sbin/mount/getmntopts.c
/*-
 * Copyright (c) 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static char sccsid[] = "@(#)getmntopts.c	8.1 (Berkeley) 3/27/94";
#endif /* not lint */

#include <sys/param.h>
#include <sys/mount.h>

#include <err.h>
#include <errno.h>
#include <fstab.h>
#include <stdlib.h>
#include <string.h>

#include "mntopts.h"

void
getmntopts(options, m0, flagp)
	const char *options;
	const struct mntopt *m0;
	int *flagp;
{
	const struct mntopt *m;
	int negative;
	char *opt, *optbuf;

	/* Copy option string, since it is about to be torn asunder... */
	if ((optbuf = strdup(options)) == NULL)
		err(1, NULL);

	for (opt = optbuf; (opt = strtok(opt, ",")) != NULL; opt = NULL) {
		char *value = strchr(opt, '=');

    	    	if (value) {
			*value++ = '\0';
		} else {
			/* Check for "no" prefix. */
			/* but only on binary options */
			if (opt[0] == 'n' && opt[1] == 'o') {
				negative = 1;
				opt += 2;
			} else
				negative = 0;
    	    	}

		/* Scan option table. */
		for (m = m0; m->m_option != NULL; ++m) {
			if (strcasecmp(opt, m->m_option) == 0)
				break;
			if (negative && strcasecmp(opt-2, m->m_option) == 0) {
				negative = 0;
				break;
			}
		}

		/* Save flag, or fail if option is not recognised. */
		if (m->m_option) {
			int *fp = m->m_flagp ? m->m_flagp : flagp;
			if (m->m_value) {
				if (!value)
					errx(1, "-o %s: requires value\n", opt);
				*m->m_value = value;
			}
			if (negative == m->m_inverse)
				*fp |= m->m_flag;
			else
				*fp &= ~m->m_flag;
		} else
			errx(1, "-o %s: option not supported", opt);
	}

	free(optbuf);
}
